magicctapipe/scripts/lst1_magic/lst1_magic_dl1_stereo_to_dl2.py

```diff
@@ -2,82 +2,266 @@
+    """
+    Reconstructs the arrival directions of shower events with the
+    MARS-like DISP method.
+
+    Parameters
+    ----------
+    event_data: pandas.core.frame.DataFrame
+        Data frame of shower events
+    tel_descriptions: dict
+        Telescope descriptions
+
+    Returns
+    -------
+    reco_params: pandas.core.frame.DataFrame
+        Data frame of the shower events with reconstructed directions
+    """
+
+    reco_params_flips = pd.DataFrame()
+
+    # First of all, we reconstruct the Alt/Az directions of all the head
+    # and tail candidates per telescope, i.e., the directions separated
+    # by the DISP parameter from the image CoG along the shower axis.
+    # Here the flip parameter (0 or 1) distinguishes the head and tail.
+
+    tel_ids = np.unique(event_data.index.get_level_values("tel_id"))
+
+    for tel_id in tel_ids:
+
+        df_events = event_data.query(f"tel_id == {tel_id}")
+
+        tel_pointing = AltAz(
+            alt=u.Quantity(df_events["pointing_alt"].to_numpy(), u.rad),
+            az=u.Quantity(df_events["pointing_az"].to_numpy(), u.rad),
+        )
+
+        tel_frame = TelescopeFrame(telescope_pointing=tel_pointing)
+
+        event_coord = SkyCoord(
+            u.Quantity(df_events["x"].to_numpy(), u.m),
+            u.Quantity(df_events["y"].to_numpy(), u.m),
+            frame=tel_descriptions[tel_id].camera.geometry.frame,
+        )
+
+        event_coord = event_coord.transform_to(tel_frame)
+
+        for flip in [0, 1]:
+
+            psi_per_flip = df_events["psi"].to_numpy() + 180 * flip
+
+            event_coord_per_flip = event_coord.directional_offset_by(
+                position_angle=u.Quantity(psi_per_flip, u.deg),
+                separation=u.Quantity(df_events["reco_disp"].to_numpy(), u.deg),
+            )
+
+            event_coord_per_flip = event_coord_per_flip.altaz
+
+            df_altaz_per_flip = pd.DataFrame(
+                data={
+                    "reco_alt": event_coord_per_flip.alt.to_value(u.deg),
+                    "reco_az": event_coord_per_flip.az.to_value(u.deg),
+                    "flip": flip,
+                },
+                index=df_events.index,
+            )
+
+            reco_params_flips = reco_params_flips.append(df_altaz_per_flip)
+
+    reco_params_flips.set_index("flip", append=True, inplace=True)
+    reco_params_flips.sort_index(inplace=True)
+
+    group_size = reco_params_flips.groupby(["obs_id", "event_id"]).size()
+    reco_params_flips["multiplicity"] = group_size
+
+    # Then, we get the flip combination minimizing the sum of the
+    # angular distances between the head and tail candidates per shower
+    # event. In order to speed up the calculations, here we separate the
+    # input events by the telescope combination types.
+
+    reco_params = pd.DataFrame()
+
+    for tel_ids in TEL_COMBINATIONS.values():
+
+        multiplicity = 2 * len(tel_ids)
```

Comment:
why do you multiply by 2 in here?

Reply:
It is because of the number of flips for each telescope event. The data frame has the multi index `[obs_id, event_id, tel_id, flip]`, so when we group per event the multiplicity becomes N_flip * N_tels.

Reply:
actually maybe it would be more clear instead of multiplying by 2 in line 159 divide by 2 in line 148 when the multiplicities are derived from group_size

Reply:
Thanks, I agree with you, so I implemented as you suggested.